File: src/org/codepulsar/ast/statements/Return.java
package org.codepulsar.ast.statements;

import org.codepulsar.ast.Expression;
import org.codepulsar.ast.Statement;

public class Return extends Statement {
    private final boolean hasExpression;
    private final Expression expression;

    private final String function;
    private final int line;

    public Return(boolean hasExpression, Expression expression, String function, int line) {
        this.hasExpression = hasExpression;
        this.expression = expression;
        this.function = function;

        this.line = line;
    }

    public <R> R accept(Visitor<R> visitor) {
        return visitor.visitReturnStatement(this);
    }

    public boolean hasExpression() {
        return this.hasExpression;
    }

    public Expression getExpression() {
        return this.expression;
    }

    public String getFunction() {
        return this.function;
    }

    public int getLine() {
        return this.line;
    }
}

File: src/codepulsar/ast/statement/Return.cpp
#include "Return.h"


Pulsar::Return::Return(Expression* value, std::string function, int line) {
    this->value = value;
    this->function = function;
    this->line = line;
}

std::any Pulsar::Return::accept(StmtVisitor& visitor) {
    return visitor.visitReturnStatement(this);
}

bool Pulsar::Return::hasValue() {
    return this->value != nullptr;
}

Pulsar::Expression* Pulsar::Return::getValue() {
    return this->value;
}

std::string Pulsar::Return::getFunction() {
    return this->function;
}

int Pulsar::Return::getLine() {
    return this->line;
}
